### PR TITLE
Ember ts

### PR DESCRIPTION
**Pull request by BarryThePenguin**
Issue: #11845 #16205

## What I did

Adds types for ember stories 

## How to test

I've included a TypeScript Story in `examples/ember-cli`

This meant I needed to add `ember-cli-typescript`. I also updated the version of `ember-cli` and friends

I'm not sure if there is anything else I should add? 👀 
Happy to make changes! 👋

### CODE REVIEW

File: app/ember/package.json [hunk 1]
```diff
@@ -43,8 +43,9 @@
   "peerDependencies": {
     "@babel/core": "*",
     "babel-plugin-ember-modules-api-polyfill": "^2.12.0",
-    "babel-plugin-htmlbars-inline-precompile": "2 || 3",
-    "ember-source": "^3.16.0"
+    "babel-plugin-htmlbars-inline-precompile": "2 || 3 || 4",
```
Comment (BarryThePenguin) [inline app/ember/package.json:46]: Based on the changelog, the only thing that has changed recently is node support

https://github.com/ember-cli/babel-plugin-htmlbars-inline-precompile/blob/master/CHANGELOG.md

File: app/ember/package.json [hunk 2]
```diff
@@ -43,8 +43,9 @@
...
+    "ember-cli-htmlbars": "5",
```
Comment (BarryThePenguin) [inline app/ember/package.json:47]: Only really needed for the `template` type returned from a story function

File: app/ember/package.json [hunk 3]
```diff
@@ -43,8 +43,9 @@
...
+    "ember-source": ">=3.16.0"
```
Comment (BarryThePenguin) [inline app/ember/package.json:48]: Unsure if this change is really necessary

File: app/ember/package.json [hunk 4]
```diff
@@ -55,6 +56,9 @@
   "gitHead": "c8cd3cc9e6b21ca651e4e3d8f39123f38a28be5c",
   "typesVersions": {
     "<3.8": {
+      "types-6-0": [
+        "ts3.4/dist/client/preview/types-6-0.d.ts"
+      ],
```
Comment (BarryThePenguin) [inline app/ember/package.json:61]: I just copied the changes from the react PR #11811

File: app/ember/src/client/preview/types.ts [hunk 5]
```diff
@@ -14,3 +17,8 @@ export interface OptionsArgs {
   context: any;
   element: any;
 }
+
+export interface StoryFnEmberReturnType<Args = DefaultArgs> {
+  template: ReturnType<typeof hbs>;
+  context: Args;
+}
```
Comment (BarryThePenguin) [inline app/ember/src/client/preview/types.ts:24]: Unlike react, ember stories just return a `{ template, context }` object, not an "element"

https://github.com/storybookjs/storybook/blob/c174d7429c2593e051cc2dd1361bb28bff71f75b/app/react/src/client/preview/types.ts

File: lib/addons/src/types.ts [hunk 6]
```diff
@@ -253,7 +253,7 @@ export interface BaseMeta<ComponentType> {
 }
 
 export interface BaseStory<Args, StoryFnReturnType> {
-  (args: Args, context: StoryContext): StoryFnReturnType;
+  (args?: Args, context?: StoryContext): StoryFnReturnType;
```
Comment (BarryThePenguin) [inline lib/addons/src/types.ts:256]: We tend to use our stories in tests too, so having `args` and `context` as required arguments for stories makes this difficult..

```ts
import { module, test } from 'qunit'
import { render } from '@ember/test-helpers'
import { GlimmerButton } from '../stories/ember-typescirpt.stories'

module('Integration | Component | button', function (hooks) {
  test('some test', async function (assert) {
	// If `args` and `context` are required here 👇 that makes tests difficult to write
    const { context, template } = GlimmerButton(/* args, context */)
    this.setProperties(context)

    await render(template)

    assert.ok('some assertion')
  })
})
```

Open to alternatives 👀

**Comment by shilman** [comment]
@gossi can you review?

File: app/ember/src/client/preview/types.ts (hunk 5 above)
Comment (gossi) [reply app/ember/src/client/preview/types.ts:24]: `context` can be any pojo. It is not `Args` only. More like: `Args & Record<string, unknown>` ?

File: app/ember/package.json (hunk 1 above)
Comment (gossi) [reply app/ember/package.json:46]: Can it work without the peer deps? If so, would prefer without them.

File: app/ember/package.json (hunk 3 above)
Comment (gossi) [reply app/ember/package.json:48]: I think, this can be removed? Ie. there won't be an ember project without this dependency. This MUST BE present on the project.

File: app/ember/package.json (hunk 2 above)
Comment (gossi) [reply app/ember/package.json:47]: This is important, although ember projects require you to have it in dependencies (or else the project is unusable). But it must be at least `>=5.2.1` (iirc 5.2 had some important change)

File: examples/ember-cli/app/templates/components/button.hbs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<button
+  {{on "click" (fn @onClick)}}
```
Comment (gossi) [inline examples/ember-cli/app/templates/components/button.hbs:2]:
```suggestion
  ...attributes
```

see below in the story then

File: examples/ember-cli/stories/ember-typescirpt.stories.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import { Story, Meta } from '@storybook/ember/types-6-0';
+import type { ButtonArgs } from '../app/components/button';
+
+export default {
+  title: 'TypeScript',
+} as Meta;
+
+const Template: Story<ButtonArgs> = (args) => ({
+  template: hbs`
+  <Button @emoji={{this.emoji}} @onClick={{this.onClick}}>
```
Comment (gossi) [inline examples/ember-cli/stories/ember-typescirpt.stories.ts:11]:
```suggestion
  <Button @emoji={{this.emoji}} {{on "click" this.onClick}}>
```

File: lib/addons/src/types.ts (hunk 6 above)
Comment (gossi) [reply lib/addons/src/types.ts:256]: `(this.)setProperties()` is a ember code that is going into extinction. While this is still working, think about alternatives, maybe:


```
this = { this, ...context };
```

I don't know if that even works 😂

Comment (gossi) [reply lib/addons/src/types.ts:256]: Never seen this kind of test writing before. Never did it myself. Interessting though 👍

**Review by gossi — COMMENTED** [review]
Cool stuff, left some comments.

File: app/ember/package.json (hunk 1 above)
Comment (BarryThePenguin) [reply app/ember/package.json:46]: I'm not entirely sure 😬 I'm happy to take direction on this 👀 

It is a dependency of `ember-cli-htmlbars`, so maybe that's enough?

File: app/ember/src/client/preview/types.ts (hunk 5 above)
Comment (BarryThePenguin) [reply app/ember/src/client/preview/types.ts:24]: `Args = DefaultArgs` which is already an indexible type. So it can be anything

https://github.com/storybookjs/storybook/blob/e940b85e5eef9eeed37402bf387e4f80d897a08c/lib/addons/src/types.ts#L35-L37

File: lib/addons/src/types.ts (hunk 6 above)
Comment (BarryThePenguin) [reply lib/addons/src/types.ts:256]: I haven't seen anything mentioning the removal of `setProperties` :eyes:

This is something provided by https://github.com/emberjs/ember-qunit#setup-tests

If it does go away, `Object.assign(this, context)` would do the job

> Never seen this kind of test writing before.

We find it pretty useful. It allows us to use our stories for testing. We can setup some story for our app in storybook, and write a test that verifies that story 🎉

File: examples/ember-cli/stories/ember-typescirpt.stories.ts (hunk 8 above)
Comment (BarryThePenguin) [reply examples/ember-cli/stories/ember-typescirpt.stories.ts:11]: The purpose of this story, and the TypeScript changes, is to show how `ButtonArgs` from the glimmer component can also be used as the `Args` type for the story.

I can make this change as the `emoji` property still shows this point

File: lib/addons/src/types.ts (hunk 6 above)
Comment (BarryThePenguin) [reply lib/addons/src/types.ts:256]: Just today, wrapped the above code into a helper function

```ts
test('some test', async function (assert) {
  await renderStory(this, GlimmerButton, { emoji: '⚽️' })
  assert.ok('some assertion')
})
```

That wraps the rendering of the story template and passing the arguments..

```js
import { render } from '@ember/test-helpers'

export async function renderStory(testContext, story, args) {
  const storyArgs = { ...story.args, ...args }
  const { context, template } = story(storyArgs)
  Object.assign(testContext, context)
  return render(template)
}
```

File: app/ember/src/client/preview/types.ts (hunk 5 above)
Comment (gossi) [reply app/ember/src/client/preview/types.ts]: not necessarily, if you pass in another type `StoryFnEmberReturnType<Foo>` and if `Foo` is an interface with defined keys, then it does not work. I think it must be something like this:

```ts
export interface StoryFnEmberReturnType<Args extends DefaultArgs = DefaultArgs> {
  // ... 
```

File: lib/addons/src/types.ts (hunk 6 above)
Comment (gossi) [reply lib/addons/src/types.ts:256]: this is neat 👍 

well `this.set` and `this.get` are transferred over from classic/legacy ember. You don't need them any longer (at least for initial values). I think, changing this post-rendering still requires `this.set`. At least personally I like to get rid of 😉

File: app/ember/package.json (hunk 1 above)
Comment (gossi) [reply app/ember/package.json]: For me that's enough 😅

File: examples/ember-cli/package.json [hunk 9]
```diff
@@ -30,21 +32,23 @@
...
+    "ember-cli-htmlbars": "^5.2.0",
     "ember-cli-inject-live-reload": "^2.0.1",
-    "ember-cli-shims": "^1.2.0",
     "ember-cli-sri": "^2.1.0",
+    "ember-cli-typescript": "^4.0.0",
+    "ember-cli-typescript-blueprints": "^3.0.0",
     "ember-cli-uglify": "^3.0.0",
     "ember-load-initializers": "^2.0.0",
-    "ember-resolver": "^7.0.0",
-    "ember-source": "~3.19.0",
```
Comment (dbendaou) [inline examples/ember-cli/package.json:47]: Would rather have extract thoses updates in a dedicated PR
